controllers/provisioners/eksfargate/eksfargate.go

```diff
@@ -0,0 +1,378 @@
+	FiniteDeletedString            = "FiniteDeleted"
+	UpdateRecoverableErrorString   = "UpdateRecoverableError"
+	UnrecoverableErrorString       = "UnrecoverableError"
+	UnrecoverableDeleteErrorString = "UnrecoverableDeleteError"
+)
+
+func New(instanceGroup *v1alpha1.InstanceGroup, worker *aws.AwsFargateWorker) (*InstanceGroupContext, error) {
+	worker.RetryLimit = 15
+	ctx := InstanceGroupContext{
+		InstanceGroup:    instanceGroup,
+		AwsFargateWorker: worker,
+	}
+	instanceGroup.SetState(v1alpha1.ReconcileInit)
+	return &ctx, nil
+}
+func CreateFargateTags(tagArray []map[string]string) map[string]*string {
+	tags := make(map[string]*string)
+	for _, t := range tagArray {
+		for k, v := range t {
+			va := new(string)
+			*va = v
+			tags[k] = va
+		}
+	}
+	return tags
+}
+
+// Convienence function to convert from json to API.
+func CreateFargateSelectors(selectors []*v1alpha1.EKSFargateSelectors) []*eks.FargateProfileSelector {
+	var eksSelectors []*eks.FargateProfileSelector
+	for _, selector := range selectors {
+		m := make(map[string]*string)
+		for k, v := range selector.Labels {
+			vv := new(string)
+			*vv = v
+			m[k] = vv
+		}
+		eksSelectors = append(eksSelectors, &eks.FargateProfileSelector{Namespace: selector.Namespace, Labels: m})
+	}
+	return eksSelectors
+}
+
+func (ctx *InstanceGroupContext) Create() error {
+	log.SetFormatter(&logrus.TextFormatter{
+		FullTimestamp: true,
+	})
+	// See if any profiles are being deleted?
+	createable, err := ctx.CanCreateAndDelete()
+	if err != nil {
+		return err
+	}
+	// Can't create the profile since other profiles are deleting.
+	if !createable {
+		return nil
+	}
+
+	instanceGroup := ctx.GetInstanceGroup()
+	var arn *string
+	if *instanceGroup.Spec.EKSFargateSpec.GetPodExecutionRoleArn() == "" {
+		arn, err = ctx.AwsFargateWorker.CreateDefaultRolePolicy()
+		if err != nil {
+			log.Errorf("Creation of default role policy failed: %v", err)
+			return err
+		}
+		// Save the role name of the default role we created.
+		ctx.GetInstanceGroup().Status.SetFargateRoleName(*ctx.AwsFargateWorker.RoleName)
+		log.Infof("Create() - Generated roleName: %s", *ctx.AwsFargateWorker.RoleName)
+	} else {
+		arn = instanceGroup.Spec.EKSFargateSpec.GetPodExecutionRoleArn()
+	}
+	log.Infof("Create() - Creating profile with arn: %s", *arn)
+	err = ctx.AwsFargateWorker.CreateProfile(arn)
+	if err != nil {
+		log.Errorf("Creation of the fargate profile for cluster %v and name %v failed: %v",
+			instanceGroup.Spec.EKSFargateSpec.GetClusterName(),
+			instanceGroup.Spec.EKSFargateSpec.GetProfileName(),
+			err)
+	} else {
+		instanceGroup.SetState(v1alpha1.ReconcileModifying)
+	}
+	return err
+}
+func (ctx *InstanceGroupContext) CloudDiscovery() error {
+	return nil
+}
+func (ctx *InstanceGroupContext) Delete() error {
+	var err error
+	worker := ctx.AwsFargateWorker
+	instanceGroup := ctx.GetInstanceGroup()
+	if instanceGroup.Status.GetFargateRoleName() != "" {
+		err = worker.DeleteDefaultRolePolicy()
```

Comment:
I think it is ok.  Status.SetFargateRoleName() is only called if we create the role.

Reply:
Oh, shouldn't we update the status with the role in use regardless of whether we created it or not? I think we should always update status, and regarding the role - we should explore some mechanism such as tags to indicate ownership (not sure if its possible to tag a role), and then only delete if it's owned by controller.

Reply:
On the request, the user specifies the Role ARN.  When the ARN is not present, I create the Role and save the Role Name in the Status.  I need the Name in order to delete the Role.  Since the user is interested in ARNs, I could save the ARN (the user-specified or generated one) in the status.  Then if the user-specified ARN is different than the ARN in the Status, then I can pull the Role name from the ARN and delete it.  Just a bit more code.